Web Inspector: replace getCSSCanvasContext with canvas

pgorszkowski-igalia opened the pull request:
#### c5ae28e2dee482dadf47c543f67e8f586aa81cbc
<pre>
Web Inspector: replace getCSSCanvasContext with canvas
<a href="https://bugs.webkit.org/show_bug.cgi?id=253414">https://bugs.webkit.org/show_bug.cgi?id=253414</a>

Reviewed by Patrick Angle and Devin Rousso.

The fix replaces unsupported on all browser function getCSSCanvasContext
with additional &quot;canvas&quot; element. This solves problem with missing popup
dialog, e.g.: to display details about hovered variable.

* Source/WebInspectorUI/UserInterface/Views/Popover.js:
(WI.Popover):

Canonical link: <a href="https://commits.webkit.org/261966@main">https://commits.webkit.org/261966@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24541227aa3a2519a05b102f8f6a059fec74af7b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111672 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3496 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120408 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22163 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11888 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3234 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117438 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16458 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99606 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105694 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98410 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/185 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46185 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13287 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/182 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/95314 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13777 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9628 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19222 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52180 "Passed tests") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8344 "Build is in progress. Recent messages:OS: Monterey (12.6.3), Xcode: 13.4.1; Checked out pull request; Reviewed by Devin Rousso and Patrick Angle; Running compile-webkit") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15767 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (changes requested):
Wouldn't this line stop working: https://searchfox.org/wubkat/rev/5b368793a8c0a3d99c6991fcd3ef96e3dbd2cf2a/Source/WebInspectorUI/UserInterface/Views/Popover.css#38 ?

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/Popover.js:
```diff
@@ -49,6 +49,14 @@ WI.Popover = class Popover extends WI.Object
+        this._canvas.style.position = "absolute";
+        this._canvas.style.top = 0;
+        this._canvas.style.left = 0;
+        this._canvas.style.zIndex = -1;
```
We should make sure style continues to be defined in CSS, so instead the `_canvas` element should get a class like `background-canvas` and have these style bits come from CSS for `.popover.background-canvas`.

patrickangle reviewed (approved):
Please remove the `[GTK][WPE]` and `Remote` portions of the commit title since this impacts all platforms and both local and remote inspection. That will make it less likely for someone to ignore this commit that affect all platforms if we have to track down any regressions caused by this.

As an aside, I wonder if we are actually doing anything so special in the background of a popover that we couldn't just set an appropriate style on the background of the popover. But that's a separate issue 😅

nt1m commented on Source/WebInspectorUI/UserInterface/Views/Popover.js:
```diff
@@ -49,6 +49,14 @@ WI.Popover = class Popover extends WI.Object
+        this._canvas = document.createElement("canvas");
+        this._canvas.style.position = "absolute";
+        this._canvas.style.top = 0;
+        this._canvas.style.left = 0;
```
I'd use `.inset = 0;`, it also sets bottom/right to 0, so this makes sure to never overflow (even though the canvas height/width are supposed to ensure that).

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/Popover.js:
```diff
@@ -49,6 +49,14 @@ WI.Popover = class Popover extends WI.Object
         this._drawBackgroundAnimationIdentifier = undefined;
+
+        // Background canvas
+        this._canvas = document.createElement("canvas");
```
this will cause us to create a `<canvas>` each time we create a `WI.Popover`, whereas with `-webkit-canvas` there was only ever one.  in order to preserve the same behavior we need to make this shared among all `WI.Popover` (e.g. make some `static _getCanvas(width, height)` or something, maybe even using a `WeakRef` so that it's not kept alive longer than it needs to be if the `<canvas>` is removed from the DOM and all `WI.Popover` are destroyed)

nt1m replied:
> this will cause us to create a <canvas> each time we create a WI.Popover

@dcrousso Isn't that expected if multiple popovers are shown at the same time (if that's possible)? Otherwise, only one of them will have a visible background

patrickangle replied:
> > this will cause us to create a  each time we create a WI.Popover
> 
> @dcrousso Isn't that expected if multiple popovers are shown at the same time (if that's possible)? Otherwise, only one of them will have a visible background

We should only ever show a single popover, matching the Cocoa platform paradigm for popovers. Anything that causes multiple popovers to appear simultaneously would be a bug, and also wouldn't work today with the shared canvas via the webkit-prefixed CSS property.

dcrousso replied:
yeah, it's not so much the concern about there being more than one `WI.Popover` as creating a _new_ `<canvas>` for _each_ `WI.Popover` (whereas with `-webkit-canvas` there was only every a _single_ `<canvas>` that was shared by _all_ `WI.Popover`)

even if every `WI.Popover` is super tiny (e.g. showing the value of a CSS variable), that's still potentially a _lot_ of `<canvas>` we'd be creating and destroying, which is probably not super great churn

pgorszkowski-igalia replied:
I have checked the "single popover" case and it seems that it is possible to have multiple popovers in web inspector. Example: 
1. set any breakpoint
2. stop in that breakpoint
3. edit that breakpoint(BreakpointPopover object is used for that)
4. move mouse cursor over some variable to display information about it(Popover object is used for that)
In such case both popovers are displayed in the same time.
I did a screenshot of that situation:
![Screenshot from 2023-03-07 14-10-56](https://user-images.githubusercontent.com/112583689/223432641-43bf1169-c00c-4847-842b-eace1d3555d5.png)

It is something separate but I describe it here just for the record(probably .

pgorszkowski-igalia replied:
> yeah, it's not so much the concern about there being more than one WI.Popover as creating a new <canvas> for each WI.Popover (whereas with -webkit-canvas there was only every a single <canvas> that was shared by all WI.Popover)
> 
> even if every WI.Popover is super tiny (e.g. showing the value of a CSS variable), that's still potentially a lot of <canvas> we'd be creating and destroying, which is probably not super great churn


Regarding to the problem "replace getCSSCanvasContext with canvas": my proposal is to create a canvas element (and add it to body element) when the first "popover" is created. When the popover is displayed "this canvas" element can be moved to popover element, and when popover is hidden then move back to body element(to have it all the time and not create each time). What do you think? I am preparing a change with this solution.

dcrousso replied:
oof that's no bueno.  we probably shouldn't allow for that 😅  @patrickangle can you file a bug to fix that?  my guess is that it's cause we only listen for `"mousedown"` to dismiss the `WI.Popover`, whereas the `WI.CodeMirrorTokenTrackingController` looks at `"mousemove"` (or something like that)

@pgorszkowski-igalia i would prefer that we don't keep the `<canvas>` alive by adding it to the `<body>` when there are no `WI.Popover` that actively need it.  how about something like this instead?
```
static _getCanvasContext(width, height)
{
    let context = WI.Popover._canvasContext?.deref();
    if (!context) {
        context = document.createElement("canvas").getContext("2d");
        WI.Popover._canvasContext = new WeakRef(context);
    }

    context.canvas.width = width;
    context.canvas.height = height;
    return context;
}   
```
which could then be used like so
```
let context = WI.Popover._getCanvasContext(scaledWidth, scaledHeight);
this._element.appendChild(context.canvas);
```
tho we may wanna do this with different timing so that we're not calling `WI.Popover._getCanvasContext` or appending the `<canvas>` multiple times for the same `WI.Popover` 🤔

pgorszkowski-igalia replied:
I filed a bug about multiple popovers: https://bugs.webkit.org/show_bug.cgi?id=253557

nt1m commented on Source/WebInspectorUI/UserInterface/Views/Popover.css:
```diff
@@ -34,8 +34,13 @@
+    position: absolute;
+    top: 0;
+    left: 0;
+    z-index : -1;
```
```suggestion
    inset: 0;
    z-index: -1;
```
I think it makes sense to set bottom/right as well to 0. `inset` should set all 4 sides.

nt1m commented on Source/WebInspectorUI/UserInterface/Views/Popover.css:
```diff
@@ -34,8 +34,13 @@
+}
 
-    background-image: -webkit-canvas(popover);
+.popover.background-canvas {
```
Do we really want to apply all the styles from `.popover` on the canvas? Should it be `.popover > .background-canvas` rather?

pgorszkowski-igalia committed:
Web Inspector: replace getCSSCanvasContext with canvas
https://bugs.webkit.org/show_bug.cgi?id=253414

Reviewed by Patrick Angle and Devin Rousso.

The fix replaces unsupported on all browser function getCSSCanvasContext
with additional "canvas" element. This solves problem with missing popup
dialog, e.g.: to display details about hovered variable.

* Source/WebInspectorUI/UserInterface/Views/Popover.js:
(WI.Popover):

Canonical link: https://commits.webkit.org/261966@main
